crypto/bio/bio_test.cc

```diff
@@ -832,21 +836,86 @@ TEST(BIOTest, ExternalData) {
 }
 
 // Test that, on Windows, |BIO_read_filename| opens files in binary mode.
-TEST(BIOTest, BinaryMode) {
+TEST(BIOTest, FileMode) {
   if (SkipTempFileTests()) {
     GTEST_SKIP();
   }
 
-  TemporaryFile file;
-  ASSERT_TRUE(file.Init("\r\n"));
+  TemporaryFile temp;
+  ASSERT_TRUE(temp.Init("hello\r\nworld"));
 
-  // Reading from the file should give back the exact bytes we put in.
+  auto expect_file_contents = [](BIO *bio, const std::string &str) {
+    // Read more than expected, to make sure we've reached the end of the file.
+    std::vector<char> buf(str.size() + 100);
+    int len = BIO_read(bio, buf.data(), static_cast<int>(buf.size()));
+    ASSERT_GT(len, 0);
+    EXPECT_EQ(Bytes(buf.data(), len), Bytes(str));
+  };
+  auto expect_binary_mode = [&](BIO *bio) {
+    expect_file_contents(bio, "hello\r\nworld");
+  };
+  auto expect_text_mode = [&](BIO *bio) {
+#if defined(OPENSSL_WINDOWS)
+    expect_file_contents(bio, "hello\nworld");
+#else
+    expect_file_contents(bio, "hello\r\nworld");
+#endif
+  };
+
+  // |BIO_read_filename| should open in binary mode.
   bssl::UniquePtr<BIO> bio(BIO_new(BIO_s_file()));
   ASSERT_TRUE(bio);
-  ASSERT_TRUE(BIO_read_filename(bio.get(), file.path().c_str()));
-  char buf[2];
-  ASSERT_EQ(2, BIO_read(bio.get(), buf, 2));
-  EXPECT_EQ(Bytes(buf, 2), Bytes("\r\n"));
+  ASSERT_TRUE(BIO_read_filename(bio.get(), temp.path().c_str()));
+  expect_binary_mode(bio.get());
+  // |BIO_new_file| does not set |BIO_FP_TEXT|, so expect it to set the file's
+  // mode to binary in all cases. This odd behavior, but it's what OpenSSL does.
+  bio.reset(BIO_new_file(temp.path().c_str(), "rb"));
+  ASSERT_TRUE(bio);
+  expect_binary_mode(bio.get());
+  bio.reset(BIO_new_file(temp.path().c_str(), "r"));
+  ASSERT_TRUE(bio);
+  // NOTE: Our behavior here aligns with OpenSSL. BoringSSL would
+  // |expect_text_mode| below because they don't call |_setmode| unless
+  // |BIO_FP_TEXT| is set.
```

Comment:
If BoringSSL does not call |_setmode| unless |BIO_FP_TEXT| is set, why does it |expect_text_mode|? Is it because it's the default? I'm a bit confused about the logic difference since |BIO_FP_TEXT| is not set. Does Boring's `BIO_new_file` set BIO_FP_TEXT when the flag is "r"?

Reply:
> Is it because it's the default?

i believe is the default file mode in Windows `fopen`, yes. please see [here](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=msvc-170#generic-text-routine-mappings):

> If `t` or `b` isn't given in mode, the default translation mode is defined by the global variable [`_fmode`](https://learn.microsoft.com/en-us/cpp/c-runtime-library/fmode?view=msvc-170).

and [here](https://learn.microsoft.com/en-us/cpp/c-runtime-library/fmode?view=msvc-170):

> The default setting of `_fmode` is `_O_TEXT` for text-mode translation.

> Does Boring's BIO_new_file set BIO_FP_TEXT when the flag is "r"?

[no](https://github.com/google/boringssl/blob/8e07cc4d884ccd2381938e309cd5c86a0c1b8cc8/crypto/bio/file.cc#L121).

`BIO_new_file` will [call `fopen`](https://github.com/google/boringssl/blob/8e07cc4d884ccd2381938e309cd5c86a0c1b8cc8/crypto/bio/file.cc#L108C10-L108C28) with whatever `mode` string is passed into `BIO_new_file`. so if the caller passes in `BIO_new_file("/some/file", "r")`, `"r"` will be interpreted natively by the underlying `fopen` implementation and the file will be opened in text mode. 

Boring will respect whatever mode the underlying file was opened with, whereas we/OpenSSL will always set it to `_O_BINARY` if `BIO_FP_TEXT` is not set in the call to `BIO_new_fp`.

This is the correctness/compatibility tradeoff described in the PR overview.

Reply:
Thank you, Will. I would suggest to clarify the comment as follows (or modify it if it's confusing)
```suggestion
  // NOTE: Our behavior here aligns with OpenSSL which is to |_setmode| the file
  // to binary. BoringSSL would |expect_text_mode| below because it respects 
  // default mode on Windows which is text and doesn't call |_setmode| (unless
  // |BIO_FP_TEXT| is set, which is not the case here).
```

Reply:
yes, that's more clear. thank you.